File: rsa_algorithm/square-and-multiply-algorithm.c
#include <stdio.h>
#include <stdint.h>
#include <math.h>

uint64_t squareAndMultiply(uint64_t M, uint64_t d, uint64_t N)
{
    uint64_t R = 1;
    uint64_t S = M;
    uint32_t n = (uint32_t)log2(d) + 1;

    for(int i = 0; i < n; i++)
    {
        if (d%2 == 1)
            R = (R * S) % N;
        S = (S * S) % N;
        d = d >> 1;
    }

    return R;
}

int main(int argc, char *argv[])
{
    uint64_t m_base = 0;
    uint64_t d_exp = 0;
    uint64_t n_mod = 0;
    uint64_t result;

    printf("Square-and-multiply algorithm for RSA\n");
    printf("Please input base size, exponent size and modulus size: ");
    scanf("%lu %lu %lu", &m_base, &d_exp, &n_mod);

    result = squareAndMultiply(m_base, d_exp, n_mod);
    printf("Result of square-and-multiply is %lu\n", result);
    return 0;
}

File: rsa_algorithm/square-and-multiply-algorithm.py
import sys
import argparse

def main():
    parser = argparse.ArgumentParser(description='Compute RSA square-and-multiply for 1024-bit keys')
    parser.add_argument('M_base', metavar='M', type=int, help='M base input parameter for square-and-multiply function')
    parser.add_argument('d_exponent', metavar='d', type=int, help='d exponent input parameter for square-and-multiply function')
    parser.add_argument('N_modulus', metavar='N', type=int, help='N modulus input parameter for square-and-multiply function')
    args = parser.parse_args()

    M = args.M_base
    d = args.d_exponent
    N = args.N_modulus

    result = squareAndMultiply(M,d,N)

    print("Result of square-and-multiply is " + str(result))

def countTotalBits(num):
    binary = bin(num)[2:]
    return (len(binary))

def squareAndMultiply(M,d,N):
    R = 1
    S = M
    n = countTotalBits(d)
    for i in range(0,n):
        if (d % 2) == 1:
            R = (R*S)%N
        S = (S*S)%N
        d = d >> 1
    return R

if __name__ == "__main__":
    main()
